112 combining spatial transforms

wyli opened the pull request:
Fixes #111 fixes #112.

### Description
implemented in pytorch for the combined spatial transforms:
- [x] affine transforms
- [x] randomised version of the transforms
- [x] random elastic deformations
- [x] unit tests
- [x] pipeline speed test (GPU/CPU)

This PR includes [a demo](https://github.com/Project-MONAI/MONAI/blob/2e1568a5751f65a42457d2a441d844cc4750922e/examples/transforms_demo_2d.ipynb) and [a speed test](https://github.com/Project-MONAI/MONAI/blob/2e1568a5751f65a42457d2a441d844cc4750922e/examples/transform_speed.ipynb).
Observations:

- the combined transforms are faster than the individual ones on both CPU/GPU
- can have a good preprocessing speed provided the output stays on GPU -- if these transforms are followed by some CPU operations, there're significant data transfer costs.
- need to run `torch.multiprocessing.set_start_method('spawn')` beforehand if both transforms and networks request GPU processing
- 3D elastic deformation field is approximated by gaussian smoothing over the coordinates because 3D cubic spline upsampling is not supported by pytorch

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

ericspod commented on monai/transforms/transforms.py:
```diff
@@ -206,10 +207,112 @@ def __call__(self, img):
+            affine = affine @ create_rotate(spatial_dims, self.rotate_params)
+        if self.shear_params:
+            affine = affine @ create_shear(spatial_dims, self.shear_params)
+        if self.translate_params:
```
it is easier to fix the order of operations rather than provide a way of determining the order, but if we do fix the order I would say shift should be last. I think it would be expected that a transform would be applied to the image as it originally is located, if a shift precedes another operation this will distort the image relative to its new position.

Nic-Ma replied:
Maybe you can support user's definition of different params in order, and construct a "Sub transform chain" based on the params order.
Thanks.

ericspod commented on monai/transforms/utils.py:
```diff
@@ -157,3 +156,108 @@ def one_hot(labels, num_classes):
+        if len(radians) >= 1:
+            affine = np.array([
+                [1., 0., 0., 0.],
+                [0., np.cos(radians[0]), -np.sin(radians[0]), 0.],
```
The cos and sin calculations each appear twice, I don't think we can expect these to be optimized together so they should be pre-calculated before this line and referenced in the array.

ericspod commented on monai/transforms/transforms.py:
```diff
@@ -206,10 +207,112 @@ def __call__(self, img):
+    transform ``img`` given the affine parameters.
+    """
+
+    def __init__(self,
```
Are the transform parameters randomly chosen at training time? I don't see how that is meant to happen quite yet.

wyli replied:
it's not in this PR yet, I'm working on it..

wyli commented:
this is ready from my side, could you review? @ericspod @atbenmurray @Nic-Ma

Nic-Ma commented on tests/test_gaussian_filter.py:
```diff
@@ -0,0 +1,58 @@
+from monai.networks.layers.simplelayers import GaussianFilter
+
+
+class IntensityNormTestCase(unittest.TestCase):
```
Typo, should not be "IntensityNorm".

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
+            a 2D (3xHxW) or 3D (4xHxWxD) grid.
+        """
+        self.randomize()
+        _affine_grid = AffineGrid(self.rotate_params, self.shear_params, self.translate_params, self.scale_params,
```
I think we should not use "_" for local variables. It's used for protected or private member variables.

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
+        return self.resampler(img, grid, mode)
+
+
+if __name__ == "__main__":
```
Suggest to remove test code from official release code.

Nic-Ma commented on monai/transforms/utils.py:
```diff
@@ -208,3 +209,136 @@ def generate_pos_neg_label_crop_centers(label, size, num_samples, pos_ratio, ran
+    raise NotImplementedError
+
+
+def create_scale(spatial_dims, s):
```
Why not use unified argument name "coefs" for all factor matrix?
I think "s" and "t" are too simple and unclear.
Thanks.

Nic-Ma commented on tests/test_random_affine.py:
```diff
@@ -0,0 +1,67 @@
+]
+
+
+class TestRandAffineGrid(unittest.TestCase):
```
Typo, should be "TestRandAffine" instead of "TestRandAffineGrid"?

Nic-Ma commented on tests/test_affine.py:
```diff
@@ -0,0 +1,64 @@
+]
+
+
+class TestResample(unittest.TestCase):
```
Typo, should be "TestAffine" instead of "TestResample"?

Nic-Ma commented on tests/test_random_elastic_3d.py:
```diff
@@ -0,0 +1,55 @@
+]
+
+
+class TestRand2DElastic(unittest.TestCase):
```
Typo, should be "TestRand3DElastic".

Nic-Ma commented on examples/transforms_demo_2d.ipynb:
```diff
@@ -0,0 +1,271 @@
+    "This demo shows how to apply 2D transforms in MONAI.\n",
+    "Main features:\n",
+    "    - Random elastic transforms implemented in native Pytorch\n",
+    "    - Easy-to-use interfaces that are designed and implmented in the pythonic way\n",
```
Typo, should be "implemented" instead of "implmented"?

Nic-Ma commented on examples/transforms_demo_2d.ipynb:
```diff
@@ -0,0 +1,271 @@
+   "metadata": {},
+   "source": [
+    "## Before running this demo\n",
+    "**please download the GLAS (gland segmentation in histology images) challenge data from\n",
```
Suggest to add ":" and "\n" at end of this line, then the notebook style will be normal.

Nic-Ma commented on examples/transform_speed.ipynb:
```diff
@@ -0,0 +1,365 @@
+    "])    \n",
+    "    \n",
+    "ds = NiftiDataset(images, segs, transform=imtrans, seg_transform=segtrans)\n",
+    "loader = DataLoader(ds, batch_size=3, num_workers=8)\n",
```
Why num_workers is 8 instead of 4 here? It's different from your other CPU test.
Thanks.

Nic-Ma commented on examples/transform_speed.ipynb:
```diff
@@ -0,0 +1,365 @@
+    "])    \n",
+    "    \n",
+    "ds = NiftiDataset(images, segs, transform=imtrans, seg_transform=segtrans)\n",
+    "loader = DataLoader(ds, batch_size=3, num_workers=4, pin_memory=torch.cuda.is_available())\n",
```
As num_workers obviously affect CPU performance and seed, is this "num_workers=4" the best value on your CPU?
And maybe you need to state your CPU type in the markdown introduction.
Thanks.

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
+        grid = create_grid(spatial_size)
+        if self.do_transform:
+            grid = torch.tensor(grid).to(self.device)
+            gaussian = GaussianFilter(3, self.sigma, 3., device=self.device)
```
As GaussianFilter contains "torch.nn.Parameter", should add "with torch.no_grad():" when run this operation?
Thanks.

wyli replied:
i set requires_grad to False https://github.com/Project-MONAI/MONAI/pull/131/files/871b074df3fada02a2d24507fe2a5ca81cddee8a?file-filters%5B%5D=.py#diff-94f123b1e6105a0be041ce6ef8461918R46

Nic-Ma replied:
OK, sounds good.

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
+        self.as_tensor_output = as_tensor_output
+        self.device = device
+
+    def __call__(self, spatial_size=None, grid=None):
```
If this class is to **generate grid** instead of **transform image**, why not define it as a utility?

wyli replied:
i think the user might load "grid" data directly and apply grid transform here

Nic-Ma replied:
OK.
1. It's acceptable.
2. But I am not sure whether we should explicitly define "transforms" only for images.
@ericspod @atbenmurray , do you guys have any comments?
Thanks.

wyli replied:
perhaps that's not needed, e.g. we allow the users to apply "seg_transforms" on classification labels as well https://github.com/Project-MONAI/MONAI/blob/master/monai/data/nifti_reader.py#L113, I think it's a nice feature for flexibilities

Nic-Ma commented on monai/transforms/transforms.py:
```diff
@@ -441,10 +444,514 @@ def __call__(self, img):
+                                      device=device)
+        self.resampler = Resample(padding_mode, as_tensor_output=as_tensor_output, device=device)
+
+    def __call__(self, img, spatial_size, mode='bilinear'):
```
Can we use the same **spatial_size** and **mode** for all images in training?
If yes, strongly suggest to move them to __init__(), then we can compose this transform with others as a transform chain.
Otherwise, I feel it's not easy to use it with other transforms together.
Thanks.

wyli commented:
> Added several comments inline.
> As this PR is very big, need to double-check the math computations carefully later.
> Thanks.

thanks! addressed most of the comments, the pr is too large to include ordering of the operations

Nic-Ma commented on monai/networks/layers/simplelayers.py:
```diff
@@ -11,6 +11,9 @@
 
 import torch
 import torch.nn as nn
+import torch.nn.functional as nnf
```
Suggest to use "F" or "f" according to almost all PyTorch official examples.
